overlord/assertstate/assertstate.go

```diff
@@ -538,6 +538,139 @@ func ValidationSetAssertionForMonitor(st *state.State, accountID, name string, s
+	deviceCtx, err := snapstate.DevicePastSeeding(st, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// refresh all currently tracked validation set assertions (this may or may not
+	// include the one requested by the caller).
+	if err = RefreshValidationSetAssertions(st, userID); err != nil {
+		return nil, err
+	}
+
+	valsets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return nil, err
+	}
+
+	getSpecificSequenceOrLatest := func(db *asserts.Database, headers map[string]string) (vs *asserts.ValidationSet, err error) {
+		var a asserts.Assertion
+		if _, ok := headers["sequence"]; ok {
+			a, err = db.Find(asserts.ValidationSetType, headers)
+		} else {
+			a, err = db.FindSequence(asserts.ValidationSetType, headers, -1, -1)
+		}
+		if err != nil {
+			return nil, err
+		}
+		vs = a.(*asserts.ValidationSet)
+		return vs, nil
+	}
+
+	// try to get existing from the db. It will be the latest one if it was
+	// tracked already and thus refreshed via RefreshValidationSetAssertions.
+	// Otherwise, it may be a local assertion that was tracked in the past and
+	// then forgotten, in which case we need to refresh it explicitly.
+	db := cachedDB(st)
+	headers := map[string]string{
+		"series":     release.Series,
+		"account-id": accountID,
+		"name":       name,
+	}
+	if sequence > 0 {
+		headers["sequence"] = fmt.Sprintf("%d", sequence)
+	}
+
+	pool := asserts.NewPool(db, maxGroups)
+	atSeq := &asserts.AtSequence{
+		Type:        asserts.ValidationSetType,
+		SequenceKey: []string{release.Series, accountID, name},
+		Sequence:    sequence,
+		Revision:    asserts.RevisionNotKnown,
+		Pinned:      sequence > 0,
+	}
+
+	vs, err = getSpecificSequenceOrLatest(db, headers)
+
+	checkForConflicts := func() error {
+		if err := valsets.Add(vs); err != nil {
+			return fmt.Errorf("internal error: cannot check validation sets conflicts: %v", err)
+		}
+		if err := valsets.Conflict(); err != nil {
+			return err
+		}
+		if err := valsets.CheckInstalledSnaps(snaps); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// found locally
+	if err == nil {
+		// check if we were tracking it already; if not, that
+		// means we found an old assertion (it was very likely tracked in the
+		// past) and we need to update it as it wasn't covered
+		// by RefreshValidationSetAssertions.
+		var tr ValidationSetTracking
+		trerr := GetValidationSet(st, accountID, name, &tr)
+		if trerr != nil && trerr != state.ErrNoState {
+			return nil, trerr
+		}
+		// not tracked, update the assertion
+		if trerr == state.ErrNoState {
+			// update with pool
+			atSeq.Sequence = vs.Sequence()
+			atSeq.Revision = vs.Revision()
+			if err := pool.AddSequenceToUpdate(atSeq, atSeq.Unique()); err != nil {
+				return nil, err
+			}
+		} else {
+			// was already tracked, add to validation sets and check
+			if err := checkForConflicts(); err != nil {
```

Comment:
I've played a bit with moving this around like so:
```
diff --git a/overlord/assertstate/assertstate.go b/overlord/assertstate/assertstate.go
index 15e12e7304..1e41bce5c9 100644
--- a/overlord/assertstate/assertstate.go
+++ b/overlord/assertstate/assertstate.go
@@ -612,6 +612,7 @@ func ValidationSetAssertionForEnforce(st *state.State, accountID, name string, s
                return nil
        }
 
+       needsUpdate := true
        // found locally
        if err == nil {
                // check if we were tracking it already; if not, that
@@ -628,25 +629,27 @@ func ValidationSetAssertionForEnforce(st *state.State, accountID, name string, s
                        // update with pool
                        atSeq.Sequence = vs.Sequence()
                        atSeq.Revision = vs.Revision()
-                       if err := pool.AddSequenceToUpdate(atSeq, atSeq.Unique()); err != nil {
-                               return nil, err
-                       }
                } else {
-                       // was already tracked, add to validation sets and check
-                       if err := checkForConflicts(); err != nil {
-                               return nil, err
-                       }
-                       return vs, nil
+                       // assertion is resolved and tracked
+                       needsUpdate = false
                }
        } else {
                if !asserts.IsNotFound(err) {
                        return nil, err
                }
+       }
 
-               // try to resolve with pool
-               if err := pool.AddUnresolvedSequence(atSeq, atSeq.Unique()); err != nil {
+       if !needsUpdate {
+               // was already tracked, add to validation sets and check
+               if err := checkForConflicts(); err != nil {
                        return nil, err
                }
+               return vs, nil
+       }
+
+       // try to resolve with pool
+       if err := pool.AddSequenceToUpdate(atSeq, atSeq.Unique()); err != nil {
+               return nil, err
        }
 
        checkBeforeCommit := func(db *asserts.Database, bs asserts.Backstore) error {
```
but the unit tests started failing, I probably missed something